### PR TITLE
Changing as_strided_scatter to deterministic inputs

### PR DESCRIPTION
**Pull request by srossross**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85583

### CODE REVIEW

**Comment by srossross** [comment]
@lezcano  @amjames 

> Usually, in OpInfos, we take care of not generating these non-deterministic inputs, because these functions are not differentiable on these inputs, and gradtests would fail on them. As such my question is, are all the functions that are marked as skips coming from non-determinism non-deterministic for all inputs, or is there any of them that's just non-deterministic for some inputs. If it's the latter, it may be the case that the function has some deeper issue as the gradtests would be failing for them.

When I update the inputs to be deterministic I would have expected the `test_fn_grad` function to be passing, is that a reasonable assumption?

I still get this error  with the input: `torch.as_strided_scatter(make_tensor([3,3]), make_tensor([2, 2]), [2, 2], [2, 1])`

```
__________________________________________________________ TestGradientsCPU.test_fn_grad_as_strided_scatter_cpu_float64 __________________________________________________________
Traceback (most recent call last):
  File "/home/srossross/pytorch/test/test_ops_gradients.py", line 156, in test_fn_grad
    self._grad_test_helper(device, dtype, op, op.get_op())
  File "/home/srossross/pytorch/test/test_ops_gradients.py", line 139, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck', check_forward_ad=check_forward_ad,
  File "/home/srossross/pytorch/test/test_ops_gradients.py", line 108, in _check_helper
    self.assertTrue(gradcheck(fn, gradcheck_args,
  File "/home/srossross/pytorch/torch/testing/_internal/common_utils.py", line 3309, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 1418, in gradcheck
    return _gradcheck_helper(**args)
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 1432, in _gradcheck_helper
    _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps,
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 1075, in _gradcheck_real_imag
    gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 1309, in _fast_gradcheck
    analytical_vJu = _get_analytical_vJu_backward_mode(inputs, outputs, nondet_tol, check_grad_dtypes, all_v, all_u_dense)
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 560, in _get_analytical_vJu_backward_mode
    all_vJ = _check_analytical_jacobian_attributes(inputs, output, nondet_tol, check_grad_dtypes,
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 549, in _check_analytical_jacobian_attributes
    raise GradcheckError('Backward is not reentrant, i.e., running backward with '
torch.autograd.gradcheck.GradcheckError: Backward is not reentrant, i.e., running backward with same input and grad_output multiple times gives different values, although analytical gradient matches numerical gradient.The tolerance for nondeterminism was 0.0.

NOTE: If your op relies on non-deterministic operations i.e., it is listed here:
https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
this failure might be expected.

If you are adding a new operator, please file an issue and then use one of the
workarounds. The workaround depends on how your test invokes gradcheck/gradgradcheck.
If the test
- manually invokes gradcheck/gradgradcheck, then call gradcheck/gradgradcheck
  with `nondet_tol=<tol>` as a keyword argument.
- is OpInfo-based (e.g., in test_ops_gradients.py), then modify the OpInfo for the test
  to have `gradcheck_nondet_tol=<tol>`.
- is a Module test (e.g., in common_nn.py), then modify the corresponding
  module_test entry to have `gradcheck_nondet_tol=<tol>`
```

I'm going to continue to investigating this

**Comment by lezcano** [comment]
Is the gradient of this function deterministic for all the inputs? (it should be if the function is differentiable)

**Comment by srossross** [comment]
> Is the gradient of this function deterministic for all the inputs? (it should be if the function is differentiable)

Just getting back to this. I don't see why it would not be since it is a non-overlapping slice operation. However since removing non-deterministic inputs does not fix any tests other than `compare_cpu` should I just close this PR?

**Comment by lezcano** [comment]
No. This PR is very valuable, because it means that the other test are real failures and should be fixed at some point. They are not just failures arising from some incorrectly chosen samples.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -274,15 +274,27 @@ def sample_inputs_as_strided_scatter(op_info, device, dtype, requires_grad, **kw
...
 
     for input_shape, output_shape, stride, storage_offset in test_cases:
         input_t = make_arg(input_shape)
         input_src = make_arg(output_shape)
         yield SampleInput(input_t, input_src, output_shape, stride, storage_offset=storage_offset)
 
+def error_inputs_as_strided_scatter(op_info, device, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=torch.float32, requires_grad=False)
+
+    input_t = make_arg([4, 4])
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:290]: Add a comment describing what error condition is being tested

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -274,15 +274,27 @@ def sample_inputs_as_strided_scatter(op_info, device, dtype, requires_grad, **kw
         ((1,), (1,), (1,), 0),
         ((3, 3), (2, 2), (1, 2), 0),
         ((3, 3), (2, 2), (1, 2), 1),
-        ((16,), (2, 2, 2, 2), (1, 1, 1, 1), 0),
-        ((16,), (2, 1, 1, 2), (1, 7, 7, 1), 0),
+        ((3, 3), (2, 2), (2, 1), 0),
+        ((16,), (2, 2, 2, 2), (8, 4, 2, 1), 0),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:278]: What's the change here? Should the test cases have a comment describing the property they assume?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -10405,14 +10418,12 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.skip('Only fails for LAZY, passes on everything else'), 'TestCompositeCompliance', 'test_backward'),  # noqa: B950
                DecorateInfo(unittest.skip('Passes on complex64 and float32 only'), 'TestJit', 'test_variant_consistency_jit'),
                DecorateInfo(unittest.skip('Fails on cuda + rocm'), 'TestCommon', 'test_complex_half_reference_testing'),
+               # https://github.com/pytorch/pytorch/issues/88105
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:10421]: Nice pointer to the issue

**Review by mruberry — COMMENTED** [review]
Overall looks good @srossross but I left some suggestions for comments to help future developers understand what's happening

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (srossross) [reply torch/testing/_internal/common_methods_invocations.py:293]: added

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (srossross) [reply torch/testing/_internal/common_methods_invocations.py:279]: added a comment above, the removed code was causing non-deterministic behaviour so I modified the strides

**Review by mruberry — APPROVED** [review]
Cool; cc @kurtamohler

**Comment by ngimel** [comment]
I believe with #88342 the backward errors should be fixed

**Comment by srossross** [comment]
I'm not sure, but could something still be wrong, because now the tests for `test_fn_gradgrad` have an unexpected success, but the `test_fn_grad_as_strided_scatter_cpu_float64` still fails with:

```
  File "/home/srossross/pytorch/torch/autograd/gradcheck.py", line 1283, in _check_analytical_numerical_equal
    raise GradcheckError(_get_notallclose_msg(a, n, j, i, complex_indices, test_imag, is_forward_ad) + jacobians_str)
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor(0., device='cuda:0', dtype=torch.float64)
analytical:tensor(1., device='cuda:0', dtype=torch.float64)

The above quantities relating the numerical and analytical jacobians are computed 
in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:

Numerical:
 tensor([[0.]], device='cuda:0', dtype=torch.float64)
Analytical:
tensor([[1.]], device='cuda:0', dtype=torch.float64)

The max per-element difference (slow mode) is: 1.0.
```

**Comment by lezcano** [comment]
I think it's fine to merge this one as is. Then we can follow up with a different PR fixing that issue. 
@bdhirsh, @soulitzer or me can have a look at it.

**Pull request closed without merging**